zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/conditional/ConditionalBehavior.java

```diff
@@ -0,0 +1,195 @@
+   *
+   * <p>Note: These filters are optional. If they are not set, apply the filter as passed.
+   *
+   * @param processDefinitionKey the process definition key to evaluate conditionals for
+   * @param variableEvents the list of variable events to evaluate conditionals for
+   */
+  public void evaluateConditionals(
+      final long processDefinitionKey, final List<VariableEvent> variableEvents) {
+
+    // performance optimization: skip evaluation if no conditional subscriptions exist for the
+    // process definition
+    if (!isConditionalSubscriptionExist(processDefinitionKey)) {
+      return;
+    }
+
+    // each subscription must be triggered only once per document merge
+    final Set<Long> triggeredSubscriptionKeys = new HashSet<>();
+    // optimize the traversal by storing interrupted scopes
+    final Set<Long> interruptedScopes = new HashSet<>();
+
+    for (final VariableEvent variableEvent : variableEvents) {
+      // start top-down traversal from the scope key where the variable was set
+      final var scopes = new ArrayDeque<>(List.of(variableEvent.scopeKey));
+
+      while (!scopes.isEmpty()) {
+        final long currentScopeKey = scopes.poll();
+
+        if (interruptedScopes.contains(currentScopeKey)) {
+          // if we reached a scope that was already interrupted, we can skip evaluating its
+          // conditionals and its children since they are also interrupted
+          break;
+        }
+
+        visitSubscriptions(
+            variableEvent, currentScopeKey, triggeredSubscriptionKeys, interruptedScopes);
+
+        if (!interruptedScopes.contains(currentScopeKey)) {
+          // not interrupted, so continue traversing to child scopes
+          elementInstanceState.forEachChild(
+              currentScopeKey, -1, (childKey, ignore) -> scopes.add(childKey));
+        }
+      }
+    }
+  }
+
+  private void visitSubscriptions(
+      final VariableEvent variableEvent,
+      final long currentScopeKey,
+      final Set<Long> triggeredSubscriptionKeys,
+      final Set<Long> interruptedScopes) {
+    conditionSubscriptionState.visitByScopeKey(
+        currentScopeKey,
+        subscription -> {
+          final var subscriptionKey = subscription.getKey();
+          final var record = subscription.getRecord();
+
+          if (triggeredSubscriptionKeys.contains(subscriptionKey)) {
+            // already triggered, so continue with next subscription
+            return true;
+          }
+
+          // apply variableNames and variableEvents filter
+          if (!variableEvent.matchesFilters(
+              record.getVariableNames(), record.getVariableEvents())) {
+            // no match, so continue with next subscription
+            return true;
+          }
+
+          if (shouldTrigger(record, currentScopeKey)) {
+            commandWriter.appendFollowUpCommand(
+                subscriptionKey, ConditionalSubscriptionIntent.TRIGGER, record);
```

Comment:
❓ Will the TRIGGER command processor be able to check whether it's supposed to actually trigger? Interruptions are tracked in memory for the duration of this behaviors execution duration. However, appending this `TRIGGER` as a follow-up command means that any scope interruption in the state is postponed. So, there may be race conditions with other commands regarding the scope's state. 

🔧 Alternatively, consider using a `TRIGGERING` event or something similar to capture this early.